cmd/bucket-handlers-listobjects.go

```diff
@@ -158,17 +134,27 @@ func (api objectAPIHandlers) ListObjectsV1Handler(w http.ResponseWriter, r *http
 	// Extract all the litsObjectsV1 query params to their native values.
 	prefix, marker, delimiter, maxKeys, _ := getListObjectsV1Args(r.URL.Query())
 
-	// Validate all the query params before beginning to serve the request.
-	if s3Error := validateListObjectsArgs(prefix, marker, delimiter, maxKeys); s3Error != ErrNone {
-		// return empty response if invalid marker
-		//TODO: avoid this pattern when moving to errors package
-		if s3Error == ErrInvalidPrefixMarker {
-			response := generateListObjectsV1Response(bucket, prefix, marker, delimiter, maxKeys, ListObjectsInfo{})
-			writeSuccessResponseXML(w, encodeResponse(response))
+	// Validate the maxKeys lowerbound. When maxKeys > 1000, S3 returns 1000 but
+	// does not throw an error.
+	if maxKeys < 0 {
+		writeErrorResponse(w, ErrInvalidMaxKeys, r.URL)
+		return
+	}
+	if !api.gateway {
+		// Validate all the query params before beginning to serve the request.
+		if s3Error := validateListObjectsArgs(prefix, marker, delimiter, maxKeys); s3Error != ErrNone {
+			writeErrorResponse(w, s3Error, r.URL)
 			return
 		}
-		writeErrorResponse(w, s3Error, r.URL)
-		return
+		// Marker is set validate pre-condition.
+		if marker != "" {
+			// Marker not common with prefix is not implemented.Send an empty response
+			if !hasPrefix(marker, prefix) {
+				response := generateListObjectsV1Response(bucket, prefix, marker, delimiter, maxKeys, ListObjectsInfo{})
+				writeSuccessResponseXML(w, encodeResponse(response))
+				return
+			}
+		}
```

Comment:
you can call validateListObjectsArgs() regardless of gateway or not. Let the underlying object layer do `if !hasPrefix(marker, prefix) {` check

Reply:
done